src/JITModule.cpp

```diff
@@ -0,0 +1,745 @@
+}
+
+void *malloc_handler(void *context, size_t x) {
+    if (context) {
+        JITUserContext *jit_user_context = (JITUserContext *)context;
+
+        return (*jit_user_context->handlers.custom_allocator.custom_malloc)(context, x);
+    } else {
+        return (*active_handlers.custom_allocator.custom_malloc)(context, x);
+    }
+}
+
+void free_handler(void *context, void *ptr) {
+    if (context) {
+        JITUserContext *jit_user_context = (JITUserContext *)context;
+
+        (*jit_user_context->handlers.custom_allocator.custom_free)(context, ptr);
+    } else {
+        (*active_handlers.custom_allocator.custom_free)(context, ptr);
+    }
+}
+
+int do_task_handler(void *context, halide_task f, int idx,
+                    uint8_t *closure) {
+    if (context) {
+        JITUserContext *jit_user_context = (JITUserContext *)context;
+
+        return (*jit_user_context->handlers.custom_do_task)(context, f, idx, closure);
+    } else {
+        return (*active_handlers.custom_do_task)(context, f, idx, closure);
+    }
+}
+
+int do_par_for_handler(void *context, halide_task f,
+                       int min, int size, uint8_t *closure) {
+    if (context) {
+        JITUserContext *jit_user_context = (JITUserContext *)context;
+
+        return (*jit_user_context->handlers.custom_do_par_for)(context, f, min, size, closure);
+    } else {
+        return (*active_handlers.custom_do_par_for)(context, f, min, size, closure);
+    }
+}
+
+void error_handler_handler(void *context, const char *msg) {
+    if (context) {
+        JITUserContext *jit_user_context = (JITUserContext *)context;
+
+        (*jit_user_context->handlers.custom_error)(context, msg);
+    } else {
+        (*active_handlers.custom_error)(context, msg);
+    }
+}
+
+int32_t trace_handler(void *context, const halide_trace_event *e) {
+    if (context) {
+        JITUserContext *jit_user_context = (JITUserContext *)context;
+
+        return (*jit_user_context->handlers.custom_trace)(context, e);
+    } else {
+        return (*active_handlers.custom_trace)(context, e);
+    }
+}
+
+template <typename function_t>
+function_t hook_function(std::map<std::string, JITModule::Symbol> exports, const char *hook_name, function_t hook) {
+    std::map<std::string, JITModule::Symbol>::const_iterator iter = exports.find(hook_name);
+    internal_assert(iter != exports.end());
+    function_t (*hook_setter)(function_t) = (function_t (*)(function_t))iter->second.address;
+
+    return (*hook_setter)(hook);
+}
+}
+
+static void adjust_module_ref_count(void *arg, int32_t count) {
+    JITModuleContents *module = (JITModuleContents *)arg;
+
+    debug(2) << "Adjusting refcount for module " << module->name << " by " << count << "\n";
+
+    if (count > 0) {
+        module->ref_count.increment();
+    } else {
+        module->ref_count.decrement();
+    }
+}
+
+namespace {
+
+#if __cplusplus > 199711L || _MSC_VER >= 1800
+std::mutex shared_runtimes_mutex;
+#endif
+
+enum RuntimeKind {
```

Comment:
Some hand-holding comments about the different types of runtime and how they're made and managed would be nice in this stretch of code.